pkg/aws/ec2/api/cleanup/eni_cleanup.go

```diff
@@ -0,0 +1,221 @@
+	}()
+	// Perform ENI cleanup after fixed time intervals till shut down variable is set to true on receiving the shutdown
+	// signal
+	for !e.shutdown {
+		e.DeleteLeakedResources()
+		time.Sleep(config.ENICleanUpInterval)
+	}
+
+	return nil
+}
+
+// DeleteLeakedResources describes all the network interfaces in available status that are created by the controller or VPC-CNI
+// This is called by periodically by ClusterENICleaner which deletes available ENIs cluster-wide, and by the NodeTermination cleaner on node termination
+// The available ENIs are deleted if ShouldDeleteENI is true, defined in the respective cleaners
+// The function also updates metrics for the periodic cleanup routine and the node termination cleanup
+func (e *ENICleaner) DeleteLeakedResources() error {
+	var errors []error
+	availableENIs := make(map[string]struct{})
+	vpcrcAvailableCount := 0
+	vpccniAvailableCount := 0
+	leakedENICount := 0
+
+	filters := CommonNetworkInterfaceFilters
+	// Append the VPC-ID deep filter for the paginated call
+	filters = append(filters, []*ec2.Filter{
+		{
+			Name:   aws.String("vpc-id"),
+			Values: []*string{aws.String(e.VpcId)},
+		},
+	}...)
+	// get cleaner specific filters
+	filters = append(filters, e.Manager.GetENITagFilters()...)
+	describeNetworkInterfaceIp := &ec2.DescribeNetworkInterfacesInput{
+		Filters: filters,
+	}
+	for {
+		describeNetworkInterfaceOp, err := e.EC2Wrapper.DescribeNetworkInterfaces(describeNetworkInterfaceIp)
+		if err != nil {
+			e.Log.Error(err, "failed to describe network interfaces, cleanup will be retried in next cycle")
+			return err
+		}
+		for _, nwInterface := range describeNetworkInterfaceOp.NetworkInterfaces {
+			if e.Manager.ShouldDeleteENI(nwInterface.NetworkInterfaceId) {
+				tagMap := utils.GetTagKeyValueMap(nwInterface.TagSet)
+				if val, ok := tagMap[config.NetworkInterfaceOwnerTagKey]; ok {
+					// Increment promethues metrics for number of leaked ENIs cleaned up
+					switch val {
+					case config.NetworkInterfaceOwnerTagValue:
+						vpcrcAvailableCount += 1
+					case config.NetworkInterfaceOwnerVPCCNITagValue:
+						vpccniAvailableCount += 1
+					default:
+						// We should not hit this case as we only filter for relevant tag values, log error and continue if unexpected ENIs found
+						e.Log.Error(fmt.Errorf("found available ENI not created by VPC-CNI/VPC-RC"), "eniID", *nwInterface.NetworkInterfaceId)
+						continue
+					}
+				}
+				_, err := e.EC2Wrapper.DeleteNetworkInterface(&ec2.DeleteNetworkInterfaceInput{
+					NetworkInterfaceId: nwInterface.NetworkInterfaceId,
+				})
+				if err != nil {
+					if !strings.Contains(err.Error(), ec2Errors.NotFoundInterfaceID) { // ignore InvalidNetworkInterfaceID.NotFound error
+						// append err and continue, we will retry deletion in the next period/reconcile
+						leakedENICount += 1
+						errors = append(errors, fmt.Errorf("failed to delete leaked network interface %v:%v", *nwInterface.NetworkInterfaceId, err))
```

Comment:
This will count eni that are being re tried too.

Reply:
Can you expand on this? In each periodic cleanup routine, this is the number of leaked ENIs that failed to be deleted. 

If these ENIs are still in leaked state, then yes, deletion will be re-tried on the next cleanup. And if it still fails deletion, it will be counted as leaked ENI which is the intended behavior. Do you have any concerns on this?

Reply:
ok I see we set this variable after every run. We should be good then. It will just count leaked eni for particular run only.